src/main/java/mate/academy/dao/BookDaoImpl.java

```diff
@@ -0,0 +1,146 @@
+            if (generatedKeys.next()) {
+                Long id = generatedKeys.getObject(FIRST_QUESTION_MARK_INDEX, Long.class);
+                book.setId(id);
+            }
+
+        } catch (SQLException e) {
+            throw new DataProcessingException(e);
+        }
+        return book;
+    }
+
+    @Override
+    public Optional<Book> findById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(TAKE_BOOK_FROM_DB)) {
+
+            statement.setLong(FIRST_QUESTION_MARK_INDEX, id);
+            ResultSet resultSet = statement.executeQuery();
+
+            if (resultSet.next()) {
+
+                Long idFromBook = resultSet.getObject(ID,
+                        Long.class);
+                String titleFromBook = resultSet.getString(TITLE);
+                BigDecimal priceFromBook = resultSet.getObject(PRICE,
+                        BigDecimal.class);
+
+                Book book = new Book(idFromBook, titleFromBook, priceFromBook);
+                return Optional.of(book);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException(e);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public List<Book> findAll() {
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(TAKE_BOOKS_FROM_DB)) {
+            ResultSet resultSet = statement.executeQuery();
+            List<Book> books = new ArrayList<>();
+            while (resultSet.next()) {
+
+                Long idFromBook = resultSet.getObject(ID,
+                        Long.class);
+                String titleFromBook = resultSet.getString(TITLE);
+                BigDecimal priceFromBook = resultSet.getObject(PRICE,
+                        BigDecimal.class);
+
+                Book book = new Book(idFromBook, titleFromBook, priceFromBook);
+                books.add(book);
+            }
+            return books;
+        } catch (SQLException e) {
+            throw new DataProcessingException(e);
+        }
+    }
+
+    @Override
+    public boolean deleteById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(DELETE_BOOK_FROM_DB)) {
+
+            statement.setLong(FIRST_QUESTION_MARK_INDEX, id);
+            statement.executeUpdate();
+            return true;
+        } catch (SQLException e) {
+            throw new DataProcessingException(e);
+        }
+    }
+
+    @Override
+    public Book update(Book book) {
+        if (book.getId() == null) {
+            throw new DataProcessingException();
+        }
+        long bookID = book.getId();
+        boolean deleteById = deleteById(bookID);
+        if (deleteById) {
+            create(book);
+        }
+        return book;
+    }
```

Comment:
It should be done like this (much more readable and using things you learned and used in other methods) for example:

```java
@Override
    public Book update(Book book) {
        String query = "UPDATE books SET title = ?, price = ? WHERE id = ?";
        try (Connection connection = ConnectionUtil.getConnection();
                PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            preparedStatement.setString(1, book.getTitle());
            preparedStatement.setBigDecimal(2, book.getPrice());
            preparedStatement.setLong(3, book.getId());
            int affectedRows = preparedStatement.executeUpdate();
            if (affectedRows < 1) {
                throw new DataProcessingException("Can't find book: "
                        + book + " - no rows in database updated");
            }
        } catch (SQLException e) {
            throw new DataProcessingException("Can't update a book: " + book, e);
        }
        return book;
    }
```